src/env/as_env.rs

```diff
@@ -1,23 +1,40 @@
 use crate::{types::Interface, GasCosts};
-use std::collections::HashMap;
+use parking_lot::RwLock;
+use std::{collections::HashMap, sync::Arc};
 use wasmer::Global;
 
 use super::Metered;
 
+/// AssemblyScript execution environment.
+///
+/// Contains the AS ffi env and all the data required to run a module.
 #[derive(Clone)]
 pub struct ASEnv {
+    /// AssemblyScript foreign function interface environment.
+    /// Used to interact with AS types.
     ffi_env: as_ffi_bindings::Env,
+    /// Set to true after a module execution was instantiated.
+    /// ABIs should be disabled in the AssemblyScript `start` function.
+    /// It prevents non-deterministic behaviour in the intances creation.
+    pub abi_enabled: Arc<RwLock<bool>>,
```

Comment:
I'm reinvoking the discussion about AtomicBool in a proper discussion. Following [this example](https://marabos.nl/atomics/atomics.html#example-stop-flag) from [Rust Atomics and Locks](https://marabos.nl/atomics/) it seems that it's the right fit for this kind of variable.

Reply:
Yes `AtomicBool` would faster but as the [doc](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html) states: 

> Note: This type is only available on platforms that support atomic loads and stores of u8.

I'm not convinced that it is the best solution but I will implement it.

Reply:
Also I am not confortable with [ordering](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html) so I will take time to understand what I'm doing before implementing anything, I believe the example you linked is too naive to be used as a reference.